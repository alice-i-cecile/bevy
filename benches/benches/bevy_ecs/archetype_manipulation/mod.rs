mod add_remove;
mod add_remove_big_sparse_set;
mod add_remove_big_table;
mod add_remove_table;
mod archetype_update;
mod commands;
mod simple_insert;
mod simple_insert_unbatched;
