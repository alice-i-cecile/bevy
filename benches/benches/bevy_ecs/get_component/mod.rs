mod get_component;
mod get_component_system;
mod world_get;
