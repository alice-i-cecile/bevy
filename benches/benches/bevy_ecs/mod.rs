use criterion::*;

mod archetype_maniplation;
mod get_component;
mod query_iteration;
mod scheduling;

fn bench_simple_insert(c: &mut Criterion) {
    let mut group = c.benchmark_group("simple_insert");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = archetype_manipulation::simple_insert::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("unbatched", |b| {
        let mut bench = archetype_manipulation::simple_insert_unbatched::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_simple_iter(c: &mut Criterion) {
    let mut group = c.benchmark_group("simple_iter");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = query_iteration::simple_iter::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("system", |b| {
        let mut bench = query_iteration::simple_iter_system::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("sparse", |b| {
        let mut bench = query_iteration::simple_iter_sparse::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("foreach", |b| {
        let mut bench = query_iteration::simple_iter_foreach::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("sparse_foreach", |b| {
        let mut bench = query_iteration::simple_iter_sparse_foreach::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_frag_iter_bc(c: &mut Criterion) {
    let mut group = c.benchmark_group("fragmented_iter");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = query_iteration::frag_iter::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("foreach", |b| {
        let mut bench = query_iteration::frag_iter_foreach::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_sparse_frag_iter(c: &mut Criterion) {
    let mut group = c.benchmark_group("sparse_fragmented_iter");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = query_iteration::sparse_frag_iter::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("foreach", |b| {
        let mut bench = query_iteration::sparse_frag_iter_foreach::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_schedule(c: &mut Criterion) {
    let mut group = c.benchmark_group("schedule");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = scheduling::schedule::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_heavy_compute(c: &mut Criterion) {
    let mut group = c.benchmark_group("heavy_compute");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = query_iteration::heavy_compute::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_add_remove(c: &mut Criterion) {
    let mut group = c.benchmark_group("add_remove_component");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("table", |b| {
        let mut bench = archetype_manipulation::add_remove_table::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("sparse_set", |b| {
        let mut bench = archetype_manipulation::add_remove_sparse_set::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_add_remove_big(c: &mut Criterion) {
    let mut group = c.benchmark_group("add_remove_component_big");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("table", |b| {
        let mut bench = archetype_manipulation::add_remove_big_table::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("sparse_set", |b| {
        let mut bench = archetype_manipulation::add_remove_big_sparse_set::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

fn bench_get_component(c: &mut Criterion) {
    let mut group = c.benchmark_group("get_component");
    group.warm_up_time(std::time::Duration::from_millis(500));
    group.measurement_time(std::time::Duration::from_secs(4));
    group.bench_function("base", |b| {
        let mut bench = get_component::get_component::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.bench_function("system", |b| {
        let mut bench = get_component::get_component_system::Benchmark::new();
        b.iter(move || bench.run());
    });
    group.finish();
}

criterion_group!(
    benchmarks,
    bench_simple_insert,
    bench_simple_iter,
    bench_frag_iter_bc,
    bench_sparse_frag_iter,
    bench_schedule,
    bench_heavy_compute,
    bench_add_remove,
    bench_add_remove_big,
    bench_get_component,
);
criterion_main!(benchmarks);
