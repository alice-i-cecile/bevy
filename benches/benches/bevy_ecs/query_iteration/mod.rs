mod frag_iter;
mod frag_iter_foreach;
mod heavy_compute;
mod simple_iter_foreach;
mod simple_iter_sparse;
mod simple_iter_sparse_foreach;
mod simple_iter_system;
mod sparse_frag_iter;
mod sparse_frag_iter_foreach;
