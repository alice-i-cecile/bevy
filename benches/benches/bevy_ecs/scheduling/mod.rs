mod run_criteria;
mod schedule;
mod stages;
