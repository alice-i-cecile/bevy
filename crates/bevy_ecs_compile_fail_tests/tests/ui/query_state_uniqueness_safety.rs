use bevy_ecs::prelude::*;

#[derive(Component, Eq, PartialEq, Debug)]
struct A(usize);

fn main() {
    let mut world = World::default();
    world.spawn().insert(A(1));

    let first_query_state: QueryState<&mut A> = world.query();
    let second_query_state: QueryState<&mut A> = world.query();

    let mut first_query = Query::from_state(&mut world, &first_query_state);
    // This should fail to compile, as another query is already active
    let mut second_query = Query::from_state(&mut world, &second_query_state);

    // This is a clear violation of no-aliased mutability
    assert_eq!(*first_query.single_mut(), *second_query.single_mut());
}
