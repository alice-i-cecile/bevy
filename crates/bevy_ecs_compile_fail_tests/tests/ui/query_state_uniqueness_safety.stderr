error[E0499]: cannot borrow `world` as mutable more than once at a time
  --> tests/ui/query_state_uniqueness_safety.rs:15:46
   |
13 |     let mut first_query = Query::from_state(&mut world, &first_query_state);
   |                                             ---------- first mutable borrow occurs here
14 |     // This should fail to compile, as another query is already active
15 |     let mut second_query = Query::from_state(&mut world, &second_query_state);
   |                                              ^^^^^^^^^^ second mutable borrow occurs here
...
18 |     assert_eq!(*first_query.single_mut(), *second_query.single_mut());
   |                 ------------------------ first borrow later used here
